File: ruby/leetcode/LongestCommonPrefix_14.rb
# 14 Longest Common Prefix

# @param {String[]} strs
# @return {String}
def longest_common_prefix(strs)
  return "" if strs.empty?

  s = ""
  base = strs[0]
  for i in 0..base.length-1
    if strs.all? { |x| x[i] == base[i] }
      s += base[i]
    else
      break
    end
  end
  s
end

strs = ["flower", "flow", "flight"]
p longest_common_prefix(strs)

File: java/problem_solving/src/leetcode/LongestCommonPrefix_14.java
package leetcode;

// #14 - Longest Common Prefix
public class LongestCommonPrefix_14 {
	public static void main(String[] args) {
		String[] strs = {"flower", "flow", "flight"};
		System.out.println(longestCommonPrefix(strs));
	}

	public static String longestCommonPrefix(String[] strs) {
		String s = strs[0];
		for (int i = 0; i < s.length(); i++) {
			String pre = s.substring(0, s.length() - i);
			for (int j = 0; j < strs.length; j++) {
				if (!strs[j].startsWith(pre)) {
					break;
				}
				if (j == strs.length - 1) {
					return pre;
				}
			}
		}
		return "";
	}
}
